Add possibility to stop motor.turn()

marvinknoll opened the pull request:
This feature adds the possibility of stopping motor.turn() before it completes.

By passing a lambda function (returning True or False) as the `stop_turn` argument, you can stop the `motor.turn()` at any point by making the lambda function return `False`.

Usually, you would want to have a thread that turns the motor. This allows you to run other operations like reading sensor data simultaneously in the main thread and then stopping the `motor.turn()` operation early.

An example usage can be found [here](https://github.com/marvinknoll/nxt-python/blob/6f83ad4ef0314da888cb582a30ce1c916409bf47/examples/tutorial/stop_turning_motor.py).

The respective mailing list discussion can be found [here](https://lists.sr.ht/~ni/nxt-python/%3CijwmxumH19YBvSyvQoUrILZoHrH6Qsg7EeUH_64Wron3ICxTWK4J8KQf4AsH5u2R56aXBe-3w-imvJ7UYcWq0IeOl0zGGXgxwLIu90MItT8%3D%40protonmail.com%3E).

schodet commented:
Thanks for your patch!

Can you find a way to avoid spinning, maybe having `time.sleep` duration limited by a maximum value when a lambda is provided?

Also, it would be nice to have the tests updated if you can.

Thanks,

Nicolas.

marvinknoll commented:
Happy to help!

I thought spinning shouldn't be a problem since no operation like `.get_tacho` is being made on the nxt. I could add a "minimal sleep" between the iterations; however, that would result in a more delayed motor stop. What do you think?

Could you please explain how to run the tests and which dependencies I need to install? Also, how have you set up your environment to use the project itself when importing it?

Many thanks,
Marvin

schodet commented:
Spinning your computer CPU is not nice for its battery or for the other tasks.

The control loop on the NXT brick runs every 100 ms, so usually you do not need faster reaction.

To run the tests, I use `python3 -m pytest`, I think you only need pytest. You can also use tox, but it is a little bit too complicated for day to day dev.

You can also run coverage with:

```
python3-coverage run
python3-coverage html
```

I open #184 to add this to a contribution guide.

schodet commented:
BTW, when running `python3 -m pytest`, it adds the current working directory to the python path, so it should use the current directory to test.

marvinknoll commented:
I just updated the function to sleep 100ms instead of spinning continuously. 

Thanks for providing the instructions on how to run the tests. I updated the failing one and added a new test for the stop motor.turn() feature.
